v2/health.go

```diff
@@ -0,0 +1,127 @@
+package shuttle
+
+import (
+	"context"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
+	"github.com/Azure/go-shuttle/v2/metrics/processor"
+	"github.com/Azure/go-shuttle/v2/metrics/sender"
+)
+
+type HealthCheckFunc func(ctx context.Context, namespace string, client *azservicebus.Client) error
+
+// HealthChecker performs periodic health checks on the Service Bus Senders and Receivers.
+// It uses azservicebus.Sender.NewMessageBatch() and azservicebus.Receiver.PeekMessages() to perform the health checks.
+type HealthChecker struct {
+	// clients is a map of namespace name to azservicebus.Client.
+	clients map[string]*azservicebus.Client
+	// entity is the name of the queue or topic.
+	entity string
+	// subscription is the name of the subscription. Leave empty for queues and senders.
+	subscription string
+	// interval is the time between health checks.
+	interval time.Duration
+	options  *HealthCheckerOptions
+}
+
+// HealthCheckerOptions configures the HealthChecker.
+// HealthCheckTimeout defaults to HealthChecker.interval if not set or set to 0 or set to be larger than interval.
+type HealthCheckerOptions struct {
+	// HealthCheckTimeout is the context timeout for each health check
+	HealthCheckTimeout time.Duration
+}
+
+// NewHealthChecker creates a new HealthChecker with the provided clients, entity, subscription, and interval.
+// clients is a map of namespace name to azservicebus.Client.
+func NewHealthChecker(clients map[string]*azservicebus.Client, entity, subscription string, interval time.Duration, options *HealthCheckerOptions) *HealthChecker {
+	if options == nil {
+		options = &HealthCheckerOptions{}
+	}
+	if options.HealthCheckTimeout == 0 || options.HealthCheckTimeout > interval {
+		options.HealthCheckTimeout = interval
+	}
+
+	return &HealthChecker{
+		clients:      clients,
+		entity:       entity,
+		subscription: subscription,
+		interval:     interval,
+		options:      options,
+	}
+}
+
+// StartSenderPeriodicHealthCheck starts a periodic health check for the sender.
+// It uses azservicebus.Sender.NewMessageBatch() to perform the health check.
+// Stops when the context is cancelled.
+func (h *HealthChecker) StartSenderPeriodicHealthCheck(ctx context.Context) {
```

Comment:
the Processor Start func blocks.

> Start starts processing on all the receivers of the processor and blocks until all processors are stopped

I think we should stay consistent and let the consumer decide to move it to a background goroutine.

Reply:
The reason why we start each client's `periodicHealthCheck`  in a goroutine is because we might have more than one namespace. If each `periodicHealthCheck` is blocking, we will never be able to move past the first namespace.